src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -76,67 +76,69 @@ public object Set(object key, object value, MemoryCacheEntryOptions cacheEntryOp
+            try
             {
-                if (cacheEntryOptions.AbsoluteExpiration <= utcNow)
+                if (_entries.TryGetValue(key, out entry))
                 {
-                    throw new ArgumentOutOfRangeException(
-                        nameof(MemoryCacheEntryOptions.AbsoluteExpiration),
-                        cacheEntryOptions.AbsoluteExpiration.Value,
-                        "The absolute expiration value must be in the future.");
+                    entry.SetExpired(EvictionReason.Removed);
                 }
+            }
+            finally
+            {
+                _entryLock.ExitReadLock();
+            }
 
-                absoluteExpiration = cacheEntryOptions.AbsoluteExpiration;
+            if (entry != null)
+            {
+                // TODO: For efficiency consider processing these removals in batches.
+                RemoveEntry(entry);
             }
 
-            var entry = new CacheEntry(
-                key,
-                value,
-                utcNow,
-                absoluteExpiration,
-                cacheEntryOptions,
-                _entryExpirationNotification);
-
-            var link = EntryLinkHelpers.ContextLink;
-            if (link != null)
-            {
-                // Copy expiration tokens and AbsoluteExpiration to the link.
-                // We do this regardless of it gets cached because the tokens are associated with the value we'll return.
-                if (entry.Options.ExpirationTokens != null)
-                {
-                    link.AddExpirationTokens(entry.Options.ExpirationTokens);
-                }
-                if (absoluteExpiration.HasValue)
+            StartScanForExpiredItems();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
                 {
-                    link.SetAbsoluteExpiration(absoluteExpiration.Value);
+                    GC.SuppressFinalize(this);
                 }
+
+                _disposed = true;
             }
+        }
 
-            bool added = false;
+        /// <inheritdoc />
+        public ICacheEntry CreateEntry(object key)
+        {
+            CheckDisposed();
+            CacheEntry priorEntry = null;
+
+            var parentEntry = CacheEntryHelper.Current;
+
+            var entry = new CacheEntry(
+                    key,
+                    SetEntry,
+                    EntryExpired
+                );
 
             _entryLock.EnterWriteLock();
             try
             {
-                if (_entries.TryGetValue(key, out priorEntry))
+                if (_entries.TryGetValue(entry.Key, out priorEntry))
                 {
-                    _entries.Remove(key);
+                    _entries.Remove(entry.Key);
                     priorEntry.SetExpired(EvictionReason.Replaced);
```

Comment:
Removing the existing entry should happen in SetEntry where you'll only need to take the write lock once.